internal/kubernetes/provider.go

```diff
@@ -61,17 +74,32 @@ func (Provider) TableName() string {
 // See https://github.com/spinnaker/clouddriver/blob/58ab154b0ec0d62772201b5b319af349498a4e3f/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/description/manifest/KubernetesKindProperties.java#L31
 // for clouddriver OSS namespace-scoped kinds.
 func (p *Provider) ValidateKindStatus(kind string) error {
-	if p.Namespace != nil {
-		for _, value := range clusterScopedKinds {
-			if strings.EqualFold(value, kind) {
-				return fmt.Errorf("namespace-scoped account not allowed to access cluster-scoped kind: '%s'", kind)
-			}
+	if p.Namespace == nil && len(p.Namespaces) == 0 {
+		return nil
+	}
+
+	for _, value := range clusterScopedKinds {
+		if strings.EqualFold(value, kind) {
+			return fmt.Errorf("namespace-scoped account not allowed to access cluster-scoped kind: '%s'", kind)
 		}
 	}
 
 	return nil
 }
 
+// ValidateNamespaceAccess verifies that this provider can access the given namespace
+func (p *Provider) ValidateNamespaceAccess(namespace string) error {
+	if namespace == "" {
+		return nil
+	}
+
+	if len(p.Namespaces) > 0 && !slice.ContainsString(p.Namespaces, namespace, nil) {
+		return fmt.Errorf("namespace-scoped account not allowed to access forbidden namespace: '%s'", namespace)
+	}
+
+	return nil
+}
```

Comment:
so you're saying if `namespace == "" && len(p.Namespaces) > 0 && !slice.ContainsString(p.Namespaces, "default", nil)` then return an error? and leave the second conditional starting on line 96 as is?

Reply:
I think this is the cleanest method. Rest of the logic looks good and should handle namespaces array validation.
```go
namespace = strings.TrimSpace(namespace)
if namespace == "" {
  namespace = "default"
}
```

Reply:
Might be a good chance to use the Go `slices.Contains()` func https://golangtutorial.dev/tips/golang-slice-contains-method/

Really we should open an issue to use that instead of any written `contains` function we have throughout the code. Generics have been available since 1.18 and it would be nice to just use built-in functions.

Reply:
So the namespace always gets validated before the kind does. So if we do it this way and we have a provider that is namespace-scoped to namespaces other than `default` and someone tries to deploy a cluster-scoped kind then the helpful error message `namespace-scoped account not allowed to access cluster-scoped kind` would get obfuscated. The namespace would get checked first, set to `default` in the validation function, and then return this error: `namespace-scoped account not allowed to access forbidden namespace: "default"`. I can imagine someone seeing this and being confused since from their perspective they weren't trying to deploy to the default namespace in the first place.

I think this is why I had it just return nil if the namespace was an empty string. I assumed that meant the kind being deployed was cluster-scoped and that `provider.ValidateKindStatus` would handle that. But that was before I knew an empty string could also mean the default namespace.

I guess I could switch the order of the checks but that would involve a bit of refactoring existing code. Or we could always allow deploying to `default`. What do you think?

Reply:
Ah OK good points.
> I guess I could switch the order of the checks but that would involve a bit of refactoring existing code.

I agree, I think this is the best approach. For `deploy.go` it would look like
```go
for _, manifest := range manifests {
  // Create a copy of the unstructured object since we access by reference.
  manifest := manifest

  err = provider.ValidateKindStatus(manifest.GetKind())
  if err != nil {
    clouddriver.Error(c, http.StatusBadRequest, err)
    return
  }
  
  err = provider.ValidateNamespaceAccess(manifest.GetNamespace()) // pass in the current manifest's namespace
  if err != nil {
    clouddriver.Error(c, http.StatusBadRequest, err)
    return
  }
  // do other stuff
 }
```
of course making note to pass in the current manifest's namespace like above. And making the changes mentioned before to change an empty namespace to default. I do see a couple tests failing, but I think those tests are validating the wrong error code.

Looks like this is everywhere the function is used.
```
./internal/api/core/search.go:165:	if err := provider.ValidateNamespaceAccess(namespace); err != nil {
./internal/api/core/kubernetes/deploy.go:81:		err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/patch.go:32:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/delete.go:34:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/cleanup.go:48:		err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/rollback.go:70:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/disable.go:48:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/restart.go:33:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/scale.go:32:	err = provider.ValidateNamespaceAccess(namespace)
./internal/api/core/kubernetes/enable.go:40:	err = provider.ValidateNamespaceAccess(namespace)
```

Reply:
Every instance of the kind and namespace checks should be in the correct order now, with the changes to the namespace validation function included